extensions/ipynb/src/notebookAttachmentCleaner.ts

```diff
@@ -0,0 +1,191 @@
+							attachmentCache[notebookUri] = { [cellFragment]: { [currFilename]: updateMetadata.custom.attachments[currFilename] } };
+						} else if (!attachmentCache[notebookUri][cellFragment]) {
+							attachmentCache[notebookUri][cellFragment] = { [currFilename]: updateMetadata.custom.attachments[currFilename] };
+						} else {
+							attachmentCache[notebookUri][cellFragment][currFilename] = updateMetadata.custom.attachments[currFilename];
+						}
+						delete updateMetadata.custom.attachments[currFilename];
+					}
+				}
+
+				// iterate second through the markdown names, pull in any cached images that we might need based on a paste, undo, etc
+				for (const currFilename of Object.keys(markdownAttachments)) {
+					// if image is addressed already --> continue, attachment will function as normal
+					if (markdownAttachments[currFilename]) {
+						continue;
+					}
+
+					// if image is referenced in mkdn && image is not in metadata -> check if image IS inside cache
+					if (attachmentCache[notebookUri][cellFragment] && Object.keys(attachmentCache[notebookUri][cellFragment]).includes(currFilename)) {
+						updateMetadata.custom.attachments[currFilename] = attachmentCache[notebookUri][cellFragment][currFilename];
+						delete attachmentCache[notebookUri][cellFragment][currFilename];
+					}
+					//TODO: ELSE: diagnostic squiggle, image not present
+
+				}
+			}
+		}
+		const metadataEdit = vscode.NotebookEdit.updateCellMetadata(currentChange.cell.index, updateMetadata);
+		const workspaceEdit = new vscode.WorkspaceEdit();
+		workspaceEdit.set(e.notebook.uri, [metadataEdit]);
+		vscode.workspace.applyEdit(workspaceEdit);
+	} // for loop of all changes
+}
+
+function cacheAllImages(metadata: { [key: string]: any }, notebookUri: string, cellFragment: string) {
+	for (const currFilename of Object.keys(metadata.custom.attachments)) {
+		if (!attachmentCache[notebookUri]) {
+			attachmentCache[notebookUri] = { [cellFragment]: { [currFilename]: metadata.custom.attachments[currFilename] } };
+		} else if (!attachmentCache[notebookUri][cellFragment]) {
+			attachmentCache[notebookUri][cellFragment] = { [currFilename]: metadata.custom.attachments[currFilename] };
+		} else {
+			attachmentCache[notebookUri][cellFragment][currFilename] = metadata.custom.attachments[currFilename];
+		}
+		delete metadata.custom.attachments[currFilename];
+	}
+}
+
+function getAttachmentNames(source: string) {
+	const filenames: any = {};
+	const re = /!\[.*?\]\(attachment:(?<filename>.*?)\)/gm;
+
+	let match;
+	while ((match = re.exec(source))) {
+		if (match.groups?.filename) {
+			filenames[match.groups?.filename] = false;
+		}
+	}
+	return filenames;
+}
+
+export function notebookAttachmentCleanerSetup(context: vscode.ExtensionContext) {
+
+	const delayTrigger = new DelayedTrigger(
+		(e) => {
+			cleanNotebookAttachments(e);
+		},
+		500
+	);
+
+	context.subscriptions.push(vscode.workspace.onDidChangeNotebookDocument(e => {
+		delayTrigger.trigger(e);
```

Comment:
If you are editing quickly, this may end up firing after `onDidCloseNotebookDocument` or `onDidRenameFiles`. Is that case properly handled?

Reply:
likely it won't be, I'm not sure exactly how I would go about handling that. Do you have any initial thoughts?

Reply:
One idea: instead of passing `e` along in a closure, add it to a list. When the delay trigger is finally invoked, it should loop through items in the list. If `close` is called before that though, remove the entry from the list

Reply:
Not completely understanding that as a solution, as when `onDidCloseNotebookDocument` is invoked, the notebook will already be closed and if a workspace edit is applied, the notebook would reopen dirty. I think this would work/there would be plenty of other solutions if an `onWillClose` listener was possible, but don't think there's any support for that at the moment.

Current fix is just to check if the notebook is closed at the very beginning of the clean function. It means that if someone is editing quickly they might not have their notebook cleaned, but not sure there is a way to hook in right before the notebook is closed.